add `is_finite` tensor element wise op and fix `is_close`/`all_close` behavior for infinity values

jonboh opened the pull request:
This pr fixes: https://github.com/tracel-ai/burn/issues/3340

There might be a better (cleaner?) way to make these operations if `is_nan`/`is_finite` are moved to the `Numeric` trait, but as `is_nan` was done at the level of `Tensor`, I implemented `is_finite` at the same level.

I've added some extra test on `test_all_close` to check these conditions.

jonboh commented:
@antimora here's the PR :)

laggui reviewed (changes requested):
Thanks for contributing 🙏 

This is certainly better than what we have in place.

While `abs(x) <= Elem::MAX` improves the robustness of the current check, I think we should have an actual `is_inf` check that compares both INF and -INF.

We don't have an associated const for infinities in the element trait, but we can still use `.elem()` or `from_elem(value)` from the `ElementConversion` trait with `f64::INFINITY` and `f64::NEG_INFINITY`. Obviously this only makes sense for float tensors, so might be best to move the implementation to float tensors only. The conversion between floating point types should work correctly.

This also made me realize that `is_close` is implemented for numeric (float AND int), but it only really makes sense for float tensors.

So overall, I think the best approach would be:
- New `is_inf` float method to check against `INFINITY` and `NEG_INFINITY`
- `is_finite` should check for (not) `is_nan` and (not) is_inf` (as your current logic)

and these methods (along with `is_close`) should belong to float only imo.

Let me know what you think!

We can do this in a follow-up PR if you prefer, but I think it makes sense to improve it here.

jonboh commented:
I agree with the changes you propose regarding the consts for INF and -INF and having the methods defined only for float tensors. I'll push the changes in this PR once I have them ready👍

antimora commented:
The code changes look good to me. 

Could you also please update the book here: https://github.com/tracel-ai/burn/blob/main/burn-book/src/building-blocks/tensor.md

New entries for is_inf and is_finite. And move is_close from numeric to float?

jonboh commented:
Ok, I've moved the ops to api/float, I've slightly modified the functions to now consume `self`. I've modified the cases in `burn-train` where this made an additional clone necessary. I think this makes sense, as the clone would otherwise be needed inside the functions themselves and that would make, in some cases, an unneeded extra clone. 
Let me know if there's anything else to improve

laggui reviewed (approved):
LGTM! Thanks again 🙂
